src/components/ControlTools.tsx

```diff
@@ -0,0 +1,68 @@
+import React, { ForwardedRef, forwardRef, useEffect } from 'react';
+import { DottingRef, useDotting } from 'dotting';
+
+import Tool from './Tool.js';
+import { FaRegTrashAlt } from 'react-icons/fa';
+import { BiRedo, BiUndo } from 'react-icons/bi';
+import { MdOutlineAddPhotoAlternate } from 'react-icons/md';
+import { TOOL_GIRD_CLASS, TOOL_TEMPLATE_CLASS } from '../styles/styleClass.js';
+
+const ControlTools = forwardRef(function InnerToolbar({}, ref: ForwardedRef<DottingRef>) {
+  // @ts-ignore
+  const { undo, redo, clear } = useDotting(ref);
+
+  useEffect(() => {
+    const keyboardUndoListener = (e: KeyboardEvent) => {
```

Comment:
From what I know, `dotting` already has an undo and redo listener!

https://github.com/hunkim98/dotting/blob/761ac38987ab1651ae7e6cc32101f0b4d43f668a/src/components/Dotting.tsx#L867-L872

If you go to the link above. You can see that there is a keydown and keyup listener in the parent div. Remind that the key down and key up event is activated when you have clicked `Dotting` canvas